### PR TITLE
fix(acp): show real rate-limit reset time and continue the turn on resume

### LINKED ISSUE #3028: Rate limit reset datetime parsed wrong
I just ran into my rate limit and aoe resolved a wrong datetime;

Additionally, after pressing `resume now` after the real reset time (which worked) aoe did not send a continuation message

### PR DESCRIPTION
**Pull request by Seluj78**
**Note:** Claude handled the implementation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

## Description

Fixes both halves of #3028.

**Wrong reset time.** claude-agent-acp's rate-limit JSON-RPC error carries only `{ "errorKind": "rate_limit" }` with no reset timestamp, so `classify_rate_limit_error` always fell back to `now + 1h`. A user in Europe/Berlin therefore saw a reset roughly an hour past the real one. The authoritative reset (`resetsAt`, a unix epoch) is forwarded out-of-band on a `usage_update`'s `_meta._claude/rateLimit`; the message text is only a locale rendering of that same value. We now capture that epoch into a per-connection atomic (cleared at each prompt start so a stale window cannot leak) and prefer it, when still in the future, over the fallback at both rate-limit classify sites. A millisecond-scale epoch is normalized back to seconds. No new dependency and no locale/timezone parsing.

**Resume did not continue.** Resuming a rate-limited session (manual "Resume now" or opt-in auto-resume) respawned the worker and restored context via `session/load` but never re-issued the interrupted prompt, so the agent sat idle. A new `EventStore::rate_limited_turn_prompt` recovers the prompt whose turn ended in `Stopped{rate_limited}`, and both resume sites enqueue it as the session's pending initial turn so the existing drain-after-live plumbing delivers it once the worker is running. It is a no-op when the last turn completed normally or was agent-initiated.

Closes #3028

## PR Type

- [x] New Feature
- [x] Bug Fix
- [ ] Refactor
- [ ] Documentation
- [ ] Infrastructure / CI

## How to test

- [ ] Rate-limit a structured-view turn (or use the fake ACP agent). Confirm the recovery banner shows the real reset time from the provider, not a time roughly one hour from now.
- [ ] While parked, click "Resume now". Confirm the agent resumes and continues the interrupted prompt instead of sitting idle.
- [ ] With `acp.rate_limit_auto_resume` enabled, let the reset window elapse. Confirm the auto-resume also re-issues the interrupted prompt.
- [ ] Rate-limit, resume, and verify a turn that had already completed normally before the limit is NOT re-sent.

## Checklist

- [x] New and existing tests pass
- [x] Documentation was updated where necessary
- [ ] For UI changes: included screenshot or recording

## Test Coverage Analysis

**Web dashboard / structured view (Playwright user story)**
- [x] Added or updated a Playwright spec under `web/tests/` and updated `web/tests/coverage-matrix.json`

**TUI / CLI (e2e test)**
- [x] N/A: this PR doesn't change TUI rendering, a CLI subcommand, or session lifecycle

## AI Usage

- [x] AI was used

**AI Model/Tool used:** Claude (via Claude Code, agent-of-empires `/aoe-implement` skill)

**Any Additional AI Details you'd like to share:**
Investigation, plan, and implementation drafted by Claude under my direction. I made the design calls (epoch capture over message parsing, reusing the pending-initial-turn drain for continuation) and reviewed each commit.

- [x] I am an AI Agent filling out this form (check box if true)

## Summary

- Rate-limit handling now uses the provider’s authoritative reset time from the `usage_update` metadata (instead of an estimated fallback), and safely ignores stale/invalid values.
- When a conversation is interrupted by a rate limit, “Resume now” now truly continues the interrupted turn by re-sending the original prompt (including attachment references).
- Auto-resume behavior was updated to queue the interrupted prompt so the resumed worker continues immediately rather than idling.
- If a newer user message arrives, any older queued rate-limit continuation is cleared to prevent replaying the wrong prompt.
- Added durable session support for queued pending-turn continuations (including attachment references) so behavior remains consistent across worker respawns/restarts.
- Expanded test coverage: added unit tests for reset-time priority and interrupted-prompt reconstruction, plus a Playwright story with a stateful fake ACP agent that persists its progress across restarts.

## Benefits

Users see the correct “rate limit reset” time and can reliably resume after rate limiting—manually or automatically—without losing or mis-replaying the interrupted work, even across restarts.

### CODE REVIEW

File: src/acp/event_store.rs [hunk 1]
```diff
@@ -1139,6 +1139,63 @@ impl EventStore {
...
+            .unwrap_or_else(|e| {
+                warn!(target: "acp.event_store", "rate_limited_turn_prompt terminator query {session_id}: {e}");
+                None
+            });
+        match terminator.and_then(|s| serde_json::from_str::<Event>(&s).ok()) {
+            Some(Event::Stopped { reason }) if reason == "rate_limited" => Some(text),
+            _ => None,
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply src/acp/event_store.rs]: Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/3061/commits/8128f26ecd5ae824139e5e5ff30021b050a5cfc8. `rate_limited_turn_prompt` now returns the interrupted prompt's attachment refs alongside its text; the pending-turn queue carries them on the instance, and the drain reloads the blobs from the event store before re-sending, so a rate-limit resume replays images/files, not just text.

File: src/server/acp_reconciler.rs [hunk 2]
```diff
@@ -1240,6 +1243,35 @@ async fn resume_one(state: Arc<AppState>, target: ResumeTarget) -> ResumeOutcome
...
+                "rate-limit continuation lookup failed: {e}"
+            );
+            return;
+        }
+    };
+    state
+        .session_service
+        .set_pending_initial_turn(id, text)
+        .await;
+}
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply src/server/acp_reconciler.rs:1275]: Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/3061/commits/8128f26ecd5ae824139e5e5ff30021b050a5cfc8. A fresh user prompt via `acp_prompt` now clears any queued `pending_initial_turn` before forwarding, so the reconciler can no longer replay an older interrupted prompt after the newer message. The drain path never routes through `acp_prompt`, so this doesn't race the continuation delivery itself.

File: src/server/api/acp.rs [hunk 3]
```diff
@@ -1179,6 +1184,12 @@ pub async fn acp_prompt(
         Ok(a) => a,
         Err((code, msg)) => return (code, msg).into_response(),
     };
+    // A fresh user prompt supersedes any queued rate-limit resume
+    // continuation, so drop it before sending: otherwise the reconciler could
+    // later replay the older interrupted prompt after this newer one (#3028).
+    // The drain path never routes through here, so this cannot race the
+    // continuation delivery itself.
+    state.session_service.clear_pending_initial_turn(&id).await;
```
Comment (jerome-benoit) [inline src/server/api/acp.rs]: [MAJOR] This clear closes the *sequential* stale-replay case, but it doesn't hold `instance_lock`. The `instance_lock` acquired at line 1095 is inside `touch_and_wake_if_sunk` and dropped when that helper returns, so at this point `acp_prompt` holds no per-session lock. Meanwhile `drain_pending_initial_turn` holds `instance_lock` across its whole snapshot -> reload -> send_turn -> clear. A fresh prompt landing while a drain is mid-await (after the drain snapshotted the old continuation) clears the field the drain already read, and the drain then sends the stale continuation *after* this newer prompt - the exact #3028 replay this guard targets. The comment above ("cannot race the continuation delivery itself") is inaccurate: it addresses handler re-entrancy, not a concurrent drain. Fix: wrap this clear + the `send_turn` below in `state.instance_lock(&id)` (the same lock the drain holds), and correct the comment.

File: web/tests/live/acp-stories/rate-limit-resume-continuation.spec.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import { test as base, expect } from "@playwright/test";
+import { spawnAoeServe, listSessions, seedSessionViaAoeAdd } from "../../helpers/aoeServe";
+import {
+  enableStructuredViewAndWait,
+  waitForStructuredView,
+  waitForReplayContains,
+  attachServeDiagnostics,
+} from "../../helpers/acp";
+
+const RESETS_AT = new Date(Date.now() + 60 * 60 * 1000).toISOString();
```
Comment (jerome-benoit) [inline web/tests/live/acp-stories/rate-limit-resume-continuation.spec.ts]: [MAJOR, tests] `RESETS_AT` is set to `now + 60min` - the very `now+1h` fallback this PR exists to replace - and the spec never asserts the displayed reset time (only the "Rate-limited" banner + the continuation marker). So the reset-time half (the PR title's first claim) has zero e2e teeth: a regression to the old guess would still pass here. Consider asserting the rendered reset time against a distinctive provider value, and driving it through a `usage_update._meta._claude/rateLimit.resetsAt` rather than the error `data.resets_at` branch.

**Review by jerome-benoit — CHANGES_REQUESTED** [review]
## Review: show real rate-limit reset time and continue the turn on resume

I verified this thoroughly, including building and running the relevant Rust suites in an isolated checkout. The core logic is correct and the unit coverage is genuinely strong, but two items should be addressed before merge - one is a real concurrency gap in the stale-replay fix, the other is a test-adequacy gap on the headline half.

### What's solid (verified)
- **Reset-time classification** is correct: the `resetsAt` epoch is captured from `usage_update._meta._claude/rateLimit`, normalized to seconds at the write site, cleared before each new turn, and preferred over the `now+1h` fallback only when strictly in the future, at both classify sites. Unit tests cover future-preferred, past-ignored, `data.resets_at` (RFC3339) beats captured, and ms->s normalization - all pass locally (8/0).
- **Resume recovery query** (`rate_limited_turn_prompt`) is correct: anchoring on the latest `UserPromptSent` and its *first* terminator, and returning `Some` only for `Stopped{rate_limited}`, correctly no-ops for normally-completed, agent-initiated-earlier, and empty-store cases (all tested, pass, 33/0).
- **CodeRabbit MAJOR #1 (attachments)** is addressed: refs ride the pending-turn flow, blobs are reloaded lazily and re-persisted idempotently (`INSERT OR IGNORE` on `(session_id, attachment_id)`), so a rate-limited turn with attachments replays without dropping or duplicating them (`rate_limited_turn_prompt_preserves_attachments` passes).
- No Phase-A/B double-fire: the parked turn reads `in_flight_turn: false`, so the continuation is a fresh prompt owned by the new connection's own future, not a control-channel adoption. CI is fully green with all heavy suites run.

### MAJOR - the stale-replay fix (CodeRabbit #2) has a concurrency gap
The fresh-prompt guard at `api/acp.rs:1192` clears `pending_initial_turn` before `send_turn`, which closes the *sequential* case. But it does not hold `instance_lock` (the `instance_lock` at `acp.rs:1095` lives inside `touch_and_wake_if_sunk` and is deliberately dropped when that helper returns), whereas `drain_pending_initial_turn` holds `instance_lock` across its whole snapshot -> reload -> `send_turn` -> clear. Since the two paths don't share the lock, a fresh user prompt landing while a drain is mid-`await` (after the drain snapshotted the old continuation) will clear the field the drain already read, and the drain then sends the stale continuation - delivering the older interrupted prompt after the newer one, which is exactly the #3028 replay the guard is meant to prevent. The in-code comment "The drain path never routes through here, so this cannot race the continuation delivery itself" is inaccurate: it reasons about handler re-entrancy, not a concurrent drain. Narrow window (the agent serializes prompts), but real and reachable on a resume immediately followed by a user prompt. Fix: wrap `acp_prompt`'s `clear_pending_initial_turn` + `send_turn` in `state.instance_lock(&id)` (the same lock the drain holds), and correct the comment.

### MAJOR (tests) - the headline reset-time half has no end-to-end coverage
The e2e (`web/tests/live/acp-stories/rate-limit-resume-continuation.spec.ts`) sets `RESETS_AT = now + 60min` (i.e. the very `now+1h` fallback the PR replaces) and never asserts the displayed reset time - so a regression of the reset-time half would still pass. Worse, the fake agent (`web/tests/helpers/fakeAcpAgent.mjs`) injects `resets_at` into the JSON-RPC error `data` (the RFC3339 branch), never into the out-of-band `usage_update._meta._claude/rateLimit.resetsAt` - so the actual #3028 mechanism (the captured-epoch atomic plus its capture/clear sites) is exercised only by Rust units, never through a live agent. Please either extend `fakeAcpAgent.mjs` to emit a `usage_update` with `_meta._claude/rateLimit.resetsAt` and assert the rendered reset time in the spec, or explicitly note that half (a) is unit-tested only.

### MINOR / NIT
- `rate_limited_turn_prompt` matches only `$.UserPromptSent`, so a `UserDiffCommentsPrompt` turn parked on `rate_limited` is not recoverable (agent stays idle for that submission type). Safe (idle, not wrong-replay), but the doc comment overstates coverage.
- The ms/s normalization threshold (`raw > 1e11`) is one-sided: a sub-1e11 milliseconds value would be misread as a far-future seconds epoch, and the `> now` filter doesn't catch a future misread. Not reachable with real near-future data; a lower-bound note would make the domain explicit.
- Auto-resume continuation and the stale-continuation-drop have no dedicated test.
- `load_attachment` returning `None` silently drops that blob via `filter_map`; a `warn!` would aid diagnosis.

Great work on the recovery query and the attachment replay - the unit coverage of the edge cases is exemplary. Happy to re-approve once the `instance_lock` gap is closed (with the comment corrected) and half (a) has some e2e teeth (or an explicit note).

File: src/server/api/acp.rs (hunk 3 above)
Comment (Seluj78) [reply src/server/api/acp.rs]: Good catch. Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/3061/commits/22acf3b888339812cf9ef7ff8b612c169d7baf81. `acp_prompt` now holds the per-session `instance_lock` across the clear + `send_turn`, the same lock the drain holds across its whole snapshot -> reload -> send -> clear, so a drain mid-await can no longer deliver the stale continuation after the newer prompt. `send_turn` doesn't take that lock (the drain already calls it while holding it), so no deadlock.

File: web/tests/live/acp-stories/rate-limit-resume-continuation.spec.ts (hunk 4 above)
Comment (Seluj78) [reply web/tests/live/acp-stories/rate-limit-resume-continuation.spec.ts]: Agreed, no teeth as written. Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/3061/commits/bb39839bcfbfcd36e6c2521f2691b55d474953ec. The reset now rides only on `usage_update._meta._claude/rateLimit.resetsAt` (the error omits `resets_at`), driving the capture path, and the spec asserts the banner renders that distinctive value (computed in-browser via the same `toLocaleTimeString`), so a regression to the now+1h guess fails.

File: src/server/api/acp.rs [hunk 5]
```diff
@@ -1179,18 +1184,31 @@ pub async fn acp_prompt(
...
+    // pending-turn drain holds that same lock across its whole snapshot ->
+    // reload -> send -> clear; without it a drain mid-await could deliver the
+    // stale continuation *after* this newer prompt. `send_turn` does not take
+    // `instance_lock` (the drain calls it while holding the lock), so this
+    // cannot deadlock. Resume + publish + forward all live in the shared
+    // service so the plugin host delivers turns through the same path (#2897).
+    let outcome = {
+        let inst_lock = state.instance_lock(&id).await;
+        let _serialized = inst_lock.lock().await;
+        state.session_service.clear_pending_initial_turn(&id).await;
```
Comment (jerome-benoit) [inline src/server/api/acp.rs:1200]: F1 confirmed closed here: the clear + `send_turn` now run under `state.instance_lock(&id)`, the same per-session lock `drain_pending_initial_turn` holds across its whole critical section (both handles share one `instance_locks` map). Atomic w.r.t. a concurrent drain, so the mid-await stale-replay window is gone. Verified no re-entrant acquisition in `send_turn`/`clear_pending_initial_turn`, so no self-deadlock, and no reverse lock ordering anywhere, so no ABBA.

File: web/tests/live/acp-stories/rate-limit-resume-continuation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+    // The turn parks on the rate limit.
+    await expect(page.getByText(/Rate-limited/i)).toBeVisible({ timeout: 15_000 });
+
+    // The banner must show the real reset from the usage_update meta, rendered
+    // the same way the UI does (`new Date(resets_at).toLocaleTimeString()`),
+    // computed in-browser so locale/timezone match. A regression to the old
+    // `now + 1h` guess would render a different time and fail this.
+    const expectedReset: string = await page.evaluate((iso) => new Date(iso).toLocaleTimeString(), RESET_ISO);
+    await expect(page.getByText(`resets at ${expectedReset}`)).toBeVisible({ timeout: 15_000 });
```
Comment (jerome-benoit) [inline web/tests/live/acp-stories/rate-limit-resume-continuation.spec.ts:104]: F2 confirmed: this asserts the rendered reset time against `now + 2h37m`, which is driven exclusively through `usage_update._meta._claude/rateLimit.resetsAt` (the error now omits `resets_at`). It cannot alias the `now + 1h` fallback, so a regression of the reset-time half would fail here. Real teeth.

**Review by jerome-benoit — APPROVED** [review]
## Re-review: rate-limit reset time + resume continuation — LGTM

Both findings from my previous review are fully addressed. I verified this the same way as last time: three independent read-only reviewers plus my own build and test run in an isolated checkout, followed by a dedicated adversarial pass that tried hard to break the APPROVE (lock-ordering deadlock, livelock, e2e triviality, incomplete fix). It held.

### F1 (stale-replay race) — RESOLVED
`22acf3b` wraps the `clear_pending_initial_turn` + `send_turn` in the per-session `instance_lock` (`api/acp.rs:1197-1211`), the same lock `drain_pending_initial_turn` holds across its whole snapshot -> reload -> send -> clear. Confirmed both handles resolve to one shared map (a single `instance_locks` is `Arc::clone`d into `SessionService` and `AppState`), so the clear+send is now atomic w.r.t. a concurrent drain, closing the mid-`await` window. No deadlock: neither `send_turn` (documents "deliberately no instance_lock here") nor `clear_pending_initial_turn` re-acquires `instance_lock`. I mapped every lock reachable under the new critical section: the `instance_lock`/`instances` domain and the supervisor `workers` domain are disjoint, and the acquisition order (`instance_lock` first, then `instances` transiently) is uniform across all call sites, so there is no ABBA reversal. The wider hold across `send_turn` merely matches the envelope the drain already imposed, so it introduces no new stall. The misleading comment was corrected too.

### F2 (reset-time had no e2e) — RESOLVED
`bb39839` removes `resets_at` from the error `data` and drives the reset time exclusively through `usage_update._meta._claude/rateLimit.resetsAt` (the real #3028 mechanism), then asserts the rendered time against a distinctive `now + 2h37m` that cannot alias the `now + 1h` fallback. Verified the fake agent forwards that meta verbatim while the error genuinely omits `resets_at`, so the assertion is gated on the captured-epoch path and would fail on a regression to the guess. No subsecond/timezone/collision flake.

### My own verification
- Build clean.
- `cargo test --features serve --lib` for the rate-limit surface: 14/14 pass (9 classify + meta, 5 `rate_limited_turn_prompt`), no hang (a deadlock in the lock fix would have hung these).
- `acp_client.rs` and `event_store.rs` are byte-identical to the previously-reviewed commit; no regression.

Nice work closing both cleanly. Minor, non-blocking: the ms/s normalization threshold is one-sided (a sub-1e11 ms value would misread), harmless with real near-future data; a lower-bound note would make the domain explicit.
